### PR TITLE
[RESTEASY-3324] Enable redirects using property.

### PR DESCRIPTION
**Pull request by The-Huginn**
https://issues.redhat.com/browse/RESTEASY-3324

### CODE REVIEW

File: resteasy-client-api/src/main/java/org/jboss/resteasy/client/jaxrs/i18n/Messages.java [hunk 1]
```diff
@@ -163,4 +163,7 @@ public interface Messages {
     @Message(id = BASE + 194, value = "No content type found in response. Cannot extract the response value.")
     @Signature(messageIndex = 1, value = { Response.class, String.class })
     ResponseProcessingException noContentTypeFound(@Param Response response);
+
+    @Message(id = BASE + 195, value = "Unable to set follow redirects policy.")
+    String unableToSetFollowRedirects();
```
Comment (jamezp) [inline resteasy-client-api/src/main/java/org/jboss/resteasy/client/jaxrs/i18n/Messages.java:168]: Given the comment below we likely don't need this.

File: resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ResteasyClientBuilderImpl.java [hunk 2]
```diff
@@ -667,6 +667,25 @@ public ResteasyClientBuilder setFollowRedirects(boolean followRedirects) {
...
+                if (enabled instanceof Boolean) {
+                    this.followRedirects = (Boolean) enabled;
+                } else if (enabled instanceof String) {
+                    this.followRedirects = Boolean.parseBoolean((String) enabled);
+                }
+            }
+        } catch (Exception e) {
+            // catch possible exceptions (in this case we do not set follow-redirects at all)
+            LogMessages.LOGGER.warn(Messages.MESSAGES.unableToSetFollowRedirects(), e);
+        }
```
Comment (jamezp) [inline resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ResteasyClientBuilderImpl.java:685]: This isn't quite right. We need to remove this whole method and check the property in the `build()` method. In the `build()` method it should just look something like:

```java
final Object localFollowRedirects = config.getProperty(PROPERTY_FOLLOW_REDIRECTS);
if (localFollowRedirects != null) {
    followRedirects = Boolean.parseBoolean(String.valueOf(localFollowRedirects));
}
```

File: resteasy-client-api/src/main/java/org/jboss/resteasy/client/jaxrs/ResteasyClientBuilder.java [hunk 3]
```diff
@@ -50,6 +50,8 @@ public enum HostnameVerificationPolicy {
      */
     public static final String PROPERTY_PROXY_SCHEME = "org.jboss.resteasy.jaxrs.client.proxy.scheme";
 
+    public static final String PROPERTY_FOLLOW_REDIRECTS = "org.jboss.resteasy.jaxrs.client.follow-redirects";
```
Comment (jamezp) [inline resteasy-client-api/src/main/java/org/jboss/resteasy/client/jaxrs/ResteasyClientBuilder.java:53]: We have started to change the prefix we use. This should be something like `dev.resteasy.client.follow.redirects`.

**Review by jamezp — CHANGES_REQUESTED** [review]
This would also need to be added to the documentation in the `RESTEasy_Client_Framework.xml` in the `docbook` directory as well.

**Comment by The-Huginn** [comment]
Currently I assume, the property has higher priority. Is that expected or do I add additional logic to allow overriding property by `setFollowRedirects()`?

**Comment by jamezp** [comment]
> Currently I assume, the property has higher priority. Is that expected or do I add additional logic to allow overriding property by `setFollowRedirects()`?

It's a tough one, but yes we are saying the property wins. If a user did something like:
```
ResteasyClientBuilder().setFollowRedirects(false)
    .property("dev.resteasy.client.follow.redirect", "true");
```

The latter would win. However, I think that's okay as then the user doesn't need to use the `ResteasyClientBuilder` directly.

**Review by jamezp — APPROVED** [review]

**Pull request merged**